server/handlers.go

```diff
@@ -814,14 +884,33 @@ func (s *Server) handleRefreshToken(w http.ResponseWriter, r *http.Request, clie
 		old.Claims.EmailVerified = ident.EmailVerified
 		old.Claims.Groups = ident.Groups
 		old.ConnectorData = ident.ConnectorData
-		old.LastUsed = s.now()
+		old.LastUsed = lastUsed
 		return old, nil
 	}
+
+	// Update LastUsed time stamp in refresh token reference object
+	// in offline session for the user.
+	if err := s.storage.UpdateOfflineSessions(refresh.Claims.UserID, refresh.ConnectorID, func(old storage.OfflineSessions) (storage.OfflineSessions, error) {
+		if old.Refresh[refresh.ClientID].ID != refresh.ID {
+			s.logger.Errorf("refresh token with id %s is invalid", refresh.ID)
+			s.tokenErrHelper(w, errInvalidRequest, "Refresh token is invalid.", http.StatusBadRequest)
```

Comment:
When you're in the middle of an update function, think of it as holding a transaction. You shouldn't wait to print or respond to web requests until you're out of the update.